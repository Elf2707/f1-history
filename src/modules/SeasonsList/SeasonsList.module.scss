.Container {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding-top: 80px;

  @media (min-width: 850px) {
    padding-top: 150px;
  }
}
