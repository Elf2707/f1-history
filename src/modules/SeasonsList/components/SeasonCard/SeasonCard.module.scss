@import '../../../../common/styles/variables/colors';

.Container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 200px;
  height: 200px;
  padding: 8px;
  margin-bottom: 16px;
  margin-left: 16px;
  box-sizing: border-box;
  box-shadow: 0 0 12px 0 $cl_black;
  background-color: $cl_light1;
  cursor: pointer;
  text-decoration: none;
  color: $cl_black;

  img {
    width: 100px;
    height: 100px;
  }

  &:hover {
    box-shadow: 0 0 45px 0 $cl_primary, inset 0 0 32px 0 $cl_primary;
  }

  &:active {
    background-color: darken($cl_light1, 10%);

    @media (min-width: 850) {
      box-shadow: 0 0 45px 0 $cl_primary, inset 0 0 32px 0 $cl_primary;
    }
  }
}
