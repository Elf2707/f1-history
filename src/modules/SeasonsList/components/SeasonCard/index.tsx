import React from 'react'
import { Link } from 'react-router-dom'

import Flag from '../../../../common/assets/images/finish-flag.png'
import styles from './SeasonCard.module.scss'

interface Props {
  year: number
}

const SeasonCard = (props: Props) => {
  return (
    <Link className={styles.Container} to={`/seasons/${props.year}`}>
      <img src={Flag} alt="image" />
      <h2>{props.year}</h2>
    </Link>
  )
}

export default SeasonCard
