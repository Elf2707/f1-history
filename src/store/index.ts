import { configure } from 'mobx'

import SeasonsStore from './seasons-store'
import Api from '../api'

export enum StoresNames {
  Seasons = 'seasons'
}

configure({ enforceActions: 'observed' })

export class Store {
  api: Api
  seasons: SeasonsStore

  constructor() {
    this.seasons = new SeasonsStore(this)
    this.api = new Api()
  }
}

export default new Store()
