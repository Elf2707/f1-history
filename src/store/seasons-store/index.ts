import { observable, computed, flow } from 'mobx'
import { FlowIterator } from 'mobx/lib/api/flow'
import _ from 'lodash'

import { Store as MainStore } from '../index'
import { GENERAL_ERROR } from '../../common/messages'
import SeasonDetails from '../../api/seasons-api/models/SeasonDetails'
import ApiResp from '../../api/models/ApiResp'
import SeasonResults from '../../api/seasons-api/models/SeasonResults'
import DriversStanding from '../../api/seasons-api/models/DriversStanding'
import Race from '../../api/seasons-api/models/Race'
import RaceResults from '../../api/seasons-api/models/RaceResults'

export default class SeasonsStore {
  @observable loading: boolean = false
  @observable seasons: number[] = _.range(2005, 2016)
  @observable error: string = ''
  @observable seasonDetails: SeasonDetails | null = null
  @observable seasonResults: SeasonResults | null = null
  @observable seasonResultsForPlace: SeasonResults | null = null
  @observable driversStandings: DriversStanding | null = null

  mainStore: MainStore

  constructor(mainStore: MainStore) {
    this.mainStore = mainStore
  }

  @computed
  get racesInSeason(): Race[] {
    if (!this.seasonResults) return []
    return this.seasonResults.RaceTable.Races.slice()
  }

  @computed
  get racesWithWinners(): RaceResults[] {
    if (!this.seasonResultsForPlace) return []
    return this.seasonResultsForPlace.RaceTable.Races.slice()
  }

  @computed
  get seasonWinner() {
    if (!this.driversStandings) return null
    return this.driversStandings.StandingsTable.StandingsLists[0].DriverStandings[0]
  }

  isGoodResponse(response: ApiResp<any>): boolean {
    return response && response.status === 200 && response.data && response.data.MRData
  }

  getSeasonDetails = flow(function*(this: SeasonsStore, seasonId: number): FlowIterator<any> {
    if (seasonId !== null && seasonId !== undefined) {
      try {
        const { api } = this.mainStore
        this.loading = true

        const response = yield api.seasonsApi.getSeasonDetails(seasonId)
        if (this.isGoodResponse(response)) {
          this.seasonDetails = response.data.MRData
          this.loading = false
        } else {
          throw new Error(GENERAL_ERROR)
        }
      } catch (error) {
        this.error = error.message || GENERAL_ERROR
        this.seasonDetails = null
        this.loading = false
      }
    }
  })

  getSeasonResults = flow(function*(this: SeasonsStore, seasonId: number): FlowIterator<any> {
    if (seasonId !== null && seasonId !== undefined) {
      try {
        const { api } = this.mainStore
        this.loading = true

        const response = yield api.seasonsApi.getSeasonResults(seasonId)
        if (this.isGoodResponse(response)) {
          this.seasonResults = response.data.MRData
          this.loading = false
        } else {
          throw new Error(GENERAL_ERROR)
        }
      } catch (error) {
        this.error = error.message || GENERAL_ERROR
        this.seasonResults = null
        this.loading = false
      }
    }
  })

  getSeasonResultsForPlace = flow(function*(
    this: SeasonsStore,
    seasonId: number,
    place: number = 1
  ): FlowIterator<any> {
    if (seasonId !== null && seasonId !== undefined) {
      try {
        const { api } = this.mainStore
        this.loading = true

        const response = yield api.seasonsApi.getSeasonResultsForPlace(seasonId, place)
        if (this.isGoodResponse(response)) {
          this.seasonResultsForPlace = response.data.MRData
          this.loading = false
        } else {
          throw new Error(GENERAL_ERROR)
        }
      } catch (error) {
        this.error = error.message || GENERAL_ERROR
        this.seasonResultsForPlace = null
        this.loading = false
      }
    }
  })

  getDriversStandings = flow(function*(this: SeasonsStore, seasonId: number): FlowIterator<any> {
    if (seasonId !== null && seasonId !== undefined) {
      try {
        const { api } = this.mainStore
        this.loading = true

        const response = yield api.seasonsApi.getDriversStandings(seasonId)
        if (this.isGoodResponse(response)) {
          this.driversStandings = response.data.MRData
          this.loading = false
        } else {
          throw new Error(GENERAL_ERROR)
        }
      } catch (error) {
        this.error = error.message || GENERAL_ERROR
        this.driversStandings = null
        this.loading = false
      }
    }
  })
}
