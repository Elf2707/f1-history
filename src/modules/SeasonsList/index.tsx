import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'

import { StoresNames } from '../../store'
import SeasonsStore from '../../store/seasons-store'
import SeasonCard from './components/SeasonCard'
import styles from './SeasonsList.module.scss'

interface Props {
  seasons: SeasonsStore
}

@inject(StoresNames.Seasons)
@observer
export default class SeasonsList extends Component<Props> {
  render() {
    return (
      <div className={styles.Container}>
        {this.props.seasons.seasons.map(season => (
          <SeasonCard key={season} year={season} />
        ))}
      </div>
    )
  }
}
