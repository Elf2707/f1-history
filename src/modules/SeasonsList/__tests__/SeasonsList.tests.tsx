import React from 'react'
import ReactDOM from 'react-dom'
import { Link, MemoryRouter } from 'react-router-dom'
import { Provider } from 'mobx-react'
import TestRenderer from 'react-test-renderer'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import SeasonsList from '../index'
import { Store } from '../../../store'
import SeasonsStore from '../../../store/seasons-store'
import SeasonCard from '../components/SeasonCard'

configure({ adapter: new Adapter() })

interface Props {
  seasonsStore: SeasonsStore
}

describe('SeasonsList Component Tests Suit', () => {
  let store: Store
  let SeasonsListView = (props: Props) => (
    <Provider {...props}>
      <MemoryRouter>
        <SeasonsList />
      </MemoryRouter>
    </Provider>
  )

  beforeEach(() => {
    store = new Store()
  })

  it('SeasonsList renders right', () => {
    const div = document.createElement('div')
    ReactDOM.render(<SeasonsListView {...store} />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('SeasonsList renders right and match snapshot', () => {
    const tree = TestRenderer.create(<SeasonsListView {...store} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('SeasonsList press on season button call navigation method', () => {
    const mockFn = jest.fn()
    const wrapper = mount(<SeasonsListView {...store} />)
    const seasonCards = wrapper.find(SeasonCard)
    const links = wrapper.find(Link)
    expect(seasonCards).toHaveLength(11)
    expect(links).toHaveLength(11)
  })
})
