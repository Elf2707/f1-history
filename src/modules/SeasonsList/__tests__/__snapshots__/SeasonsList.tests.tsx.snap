// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeasonsList Component Tests Suit SeasonsList renders right and match snapshot 1`] = `
<div
  className="Container"
>
  <a
    className="Container"
    href="/seasons/2005"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2005
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2006"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2006
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2007"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2007
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2008"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2008
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2009"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2009
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2010"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2010
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2011"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2011
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2012"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2012
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2013"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2013
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2014"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2014
    </h2>
  </a>
  <a
    className="Container"
    href="/seasons/2015"
    onClick={[Function]}
  >
    <img
      alt="image"
      src="finish-flag.png"
    />
    <h2>
      2015
    </h2>
  </a>
</div>
`;
